File: app/src/main/java/me/zhanghai/android/files/util/StatefulData.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.util;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class StatefulData<T> {

    public enum State {
        LOADING,
        ERROR,
        SUCCESS
    }

    @NonNull
    public final State state;
    @Nullable
    public final T data;
    @Nullable
    public final Exception exception;

    protected StatefulData(@NonNull State state, @Nullable T data, @Nullable Exception exception) {
        this.state = state;
        this.data = data;
        this.exception = exception;
    }
}

File: app/src/main/java/me/zhanghai/android/files/util/StatefulData.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.util

open class StatefulData<T> protected constructor(
    val state: State,
    val data: T?,
    val exception: Exception?
) {
    enum class State {
        LOADING,
        ERROR,
        SUCCESS
    }
}
